src/cata_variant.h

```diff
@@ -0,0 +1,153 @@
+#pragma once
+#ifndef CATA_VARIANT_H
+#define CATA_VARIANT_H
+
+#include <array>
+#include <cassert>
+#include <utility>
+
+#include "debug.h"
+#include "type_id.h"
+
+using itype_id = std::string;
+
+// cata_variant is a variant-like type that stores a variety of different cata
+// types.  All types are stored by converting them to a string.
+
+enum class cata_variant_type : int {
+    itype_id,
+    mtype_id,
+    num_types, // last
+};
+
+// Here follows various implementation details.  Skip to the bottom of the file
+// to see cata_variant itself.
+
+namespace cata_variant_detail
+{
+
+// Converting cata_variant_type enum values to and from string for serialization and error
+// reporting.
+std::string to_string( cata_variant_type );
+cata_variant_type from_string( const std::string & );
+
+// The convert struct is specialized for each cata_variant_type to provide the
+// code for converting that type to or from a string.
+template<cata_variant_type Type>
+struct convert;
+
+// type_for<T>() returns the cata_variant_type enum value corresponding to the
+// given value type.  e.g. type_for<mtype_id>() == cata_variant_type::mtype_id.
+
+template<typename T, size_t... I>
+constexpr cata_variant_type type_for_impl( std::index_sequence<I...> )
+{
+    constexpr size_t num_types = static_cast<size_t>( cata_variant_type::num_types );
+    constexpr std::array<bool, num_types> matches = {{
+            std::is_same<T, typename convert<static_cast<cata_variant_type>( I )>::type>::value...
+        }
+    };
+    for( size_t i = 0; i < num_types; ++i ) {
+        if( matches[i] ) {
+            return static_cast<cata_variant_type>( i );
+        }
+    }
+    // No match
+    return cata_variant_type::num_types;
+}
+
+template<typename T>
+constexpr cata_variant_type type_for()
+{
+    constexpr size_t num_types = static_cast<size_t>( cata_variant_type::num_types );
+    using SimpleT = std::remove_reference_t<T>;
+    return type_for_impl<SimpleT>( std::make_index_sequence<num_types> {} );
+}
+
+// These are the specializations of convert for each value type.
+
+template<>
+struct convert<cata_variant_type::itype_id> {
+    using type = itype_id;
```

Comment:
Is this using statement accomplishing anything?

Reply:
Yes.  It allows the lookup from a `cata_variant_type` value to the corresponding element type and vice versa.  It's used in `type_for_impl` above and `cata_variant::get`'s return type below.

I could have just used the return type from `from_string`, but that prevents the optimization discussed above for avoiding including expensive headers, and I would argue is harder to read.